src/man/latino.1
.TH "Latino" "MIT/X11" "Copyright (c) 2015-2019" "Primitivo R."
.SH Sintesis
latino [ \fIopciones\fR ] [ \fIarchivo.lat\fR | \fIcadena\fR ]
.PP
.BR Opciones
.br
-a      : Muestra la ayuda de Latino
.br
-e      : Ejecuta una cadena de codigo
.br
-v      : Muestra la version de Latino
.SH Introducción
.BR "Latino es un nuevo lenguaje de programación para el público en general."
.PP
¿Para qué aprender una nueva sintaxis?
.PP
Hemos visto varios lenguajes de programación, pero ¿Cuándo un lenguaje fácil y en español?
.br
Es cierto que ya existen desarrollados lenguajes en español, pero la idea de Latino, es crear un lenguaje
.br
totalmente fácil tanto para niños como para adultos siendo un lenguaje de programación avanzado.
.PP
¿Que tanto puedo hacer con Latino?
.br
Lo mismo que en cualquier lenguaje de programación, incluye como \fIfunciones\fR
.PP
.PP
.SH Aprendiendo a programar
.PP
.IR Incluir
.PP
En Latino se puede incluir otras librerías de Latino o librerías creadas por nosotros, como
.br
la mayoría de lenguajes de programación.
.PP
La sintaxis es la siguiente:
.RS
.BR incluir('mi_libreria')
.RE
.PP
Como ejemplo, podriamos crear un archivo llamado \fBmi_libreria.lat\fR y lo incluimos con la sintaxis anterior.
.br
Podremos agregar una \fIfuncion\fR y llamarla en otro código incluyendo la librería \fBmi_libreria\fR.
.PP
Posteriormente, podemos crear cualquier otra función o variable en la librería para
.br
así, no necesitar añadir las funciones que vamos a usar en nuestro código principal.
.PP
mi_libreria.lat:
.PP
.RS
\fB
.br
funs = {}
.br
funs.mas10 = funcion(numero)
.RS
poner(numero+10)
.RE
fin
.br
funs.por10 = funcion(numero)
.RS
poner(numero*10)
.RE
fin
.br
funs.menos10 = funcion(numero)
.RS
poner(numero-10)
.RE
fin
.br
funs.entre10 = funcion(numero)
.RS
poner(numero/10)
.RE
fin
.br
ret funs
.RE
.PP
menu.lat:
.PP
.RS
/*
.br
** Funcion Menu: obtiene los argumentos desde que se ejecuta script
.br
** ejemplo: latino script.lat 100
.br
*/
\fB
.br
mis_funs = incluir("mi_libreria")
.br
funcion menu(argc, argv)
.RS
numero_argumentos = argc \fR/* numero argumentos */\fB
.br
numero = argv[1] \fR/* primer argumento */\fB
.br
si numero_argumentos == 1 \fR/* argumento[0] = menu.lat */\fB
.RS
error("Faltan argumentos al ejecutar")
.RE
fin
.br
mis_funs.mas10(numero)
.br
mis_funs.menos10(numero)
.br
mis_funs.por10(numero)
.br
mis_funs.entre10(numero)
.RE
fin
\fR
.RE
.PP
Por último, es \fBimportante\fR recordar que si deseas usar \fBincluir("libreria")\fR sin asignarle
.br
una variable, es necesario usar funciones tipo \fBglobal\fR, en caso contrario, no se incluirán.
.PP
.IR Funciones
.PP
También, podemos crear funciones incluyendo parametros entre parentesis \fB()\fR
.PP
La sintaxis sería la siguiente:
.PP
.RS
/* Funciones locales */
\fB
.br
funcion nombre(argumentos)
.RS
\fR/* Código */\fB
.RE
fin
.PP
/* Funciones globales */
\fB
.br
global funcion nombre(argumentos)
.RS
\fR/* Código */\fB
.RE
fin
\fR
.RE
.RE
.PP
.IR Comentarios
.PP
Si queremos añadir comentarios en el código, podemos hacerlo de varias formas.
.PP
Comentarios simples:
.RS
\fB
.br
// este es un ejemplo de comentario de una línea
.br
# este es otro ejemplo de comentario
\fR
.RE
.PP
Comentarios múltiples:
.RS
\fB
.br
/*
.br
** Este es un
.br
** comentario
.br
** múltiple
.br
*/
\fR
.RE
.PP
.IR Variables
.PP
Podemos declarar una variable simple, para en sí, llamar a funciones de una forma simple.
.PP
Sintaxis de valores declarados:
.RS
\fB
.br
mi_valor="texto"
.br
mi_otro_valor = "prueba"
\fR
.RE
.PP
Como vemos, a diferencia de otras sintaxis, no importa si el valor declarado lleva el símbolo \fB=\fR entre espacios ó no.
.PP
.PP
Por defecto, el alcance de las variables es de manera local (no se incluyen, salen de condiones, etc)
.br
Pero podemos asignarlas de manera global:
.br
.RS
\fB
.br
global variable = valor
\fR
.RE
.PP
.IR Expresiones
.PP
Además, podremos crear expresiones con la impresión de texto, pondremos un ejemplo con \fBescribir()\fR
.PP
Sintaxis:
.br
.RS
.BR poner(1+2)
.RE
.br
Ó, en valores declarados:
.RS
\fB
.br
a=4
.br
suma=a+2
.br
division=a/2
.br
multip=a*2
.br
poner(suma)
\fR
.RE
.PP
Los operadores de expresión son los siguientes:
.br
.RS
.br
.BR +
.br
.BR -
.br
.BR *
.br
.BR /
.br
.BR %
.RE
.PP
.IR "Condiciones (si, osi)"
.PP
Podemos añadir condiciones en Latino (esquema de decisión condicional)
.br
En otras programaciones lo conocerás como if, else, etc.
.br
La sintaxis es la siguiente:
.PP
.RS
/* Condición: establecida por el usuario */
\fB
.br
si condición
.RS
\fR/* Código */\fB
.RE
fin
.PP
/* Condición: variable es igual a...? */
\fB
.br
si variable == "cadena"
.RS
\fR/* Código */\fB
.RE
fin
.PP
/* Condición: variable no es igual a...? */
\fB
.br
si variable != "cadena"
.RS
\fR/* Código */\fB
.RE
fin
.PP
/*
.br
** Condición: variable no es igual a..
.br
** Y
.br
** Condición: variable_2 igual a..
.br
*/
\fB
.br
si variable != "cadena" && variable_2 == 5
.RS
\fR/* Código */\fB
.RE
fin
.PP
/*
.br
** Condición: variable no es igual a..
.br
** O
.br
** Condición: variable no es igual a..
.br
*/
\fB
.br
si variable != "cadena" || variable != 5
.RS
\fR/* Código */\fB
.RE
fin
.PP
/* Podemos ponerlo con ó sin paréntesis, es opcional. */
.br
/* Aquí unos ejemplos: */
\fB
.br
si (condición)
.RS
\fR/* Código */\fB
.RE
fin
.PP
\fB
.br
si (variable == "cadena")
.RS
\fR/* Código */\fB
.RE
fin
.PP
/* Podríamos poner varias condiciones: osi */
\fB
.br
si (condición)
.RS
\fR/* Código */\fB
.RE
osi (otra_condición)
.RS
\fR/* Código */\fB
.RE
fin
\fR
.RE
.PP
.IR "Bucles (mientras, desde, repetir-hasta)"
.PP
Podemos crea un bucle, para así, no salir del script y continue siempre (condición=verdadero)
.br
ó que termine hasta cierta condición.
.PP
Sintaxis:
.RS
/* Creará un bucle con condición */
\fB
.br
mientras (condición)
.RS
\fR/* Código */\fB
.RE
fin
.PP
/* Imprimirá desde 0 a 10 */
\fB
.br
desde(i=0; i<10; i++)
.br
.RS
poner(i)
.RE
fin
.PP
\fB
.br
i = 0
.br
repetir
.br
.RS
poner(i)
.br
i++
.RE
hasta (i >= 10)
\fR
.RE
.PP
.IR "Elegir (switch)"
.PP
Una manera más limpia de condicionar variable a valor igual a, es usando
.br
la sentencia de seleccion \fBelegir\fR:
.PP
.RS
\fB
.br
mi_dinero = 10 \fR/* el valor será 10 */\fB
.br
elegir(mi_dinero)
.br
caso 1: \fR/* si valor==1 */\fB
.RS
poner("Tenemos 1 MXN")  \fR/* hacer... */\fB
.RE
caso 5: \fR/* si valor==5 */\fB
.RS
poner("Tenemos 5 MXN") \fR/* hacer...*/\fB
.RE
caso 10: \fR/* si valor==10 */\fB
.RS
poner("Tenemos 10 MXN") \fR/* hacer... */\fB
.RE
defecto: \fR/* si valor!=1 && valor!=5 && valor!=10 */\fB
.RS
poner("No sé cuanto tenemos.") \fR/* hacer... */\fB
.RE
fin
\fR
.RE
.PP
.IR "Diccionarios - Arrays (unidimensionales, bidimensionales o multidimensionales)"
.PP
También se puede en Latino hacer listas, arrays, diccionarios o como los conozcas:
.PP
.RS
\fB
.br
datos = {
.RS
"gane" : "100 MXN",
.br
"gaste" : "90 MXN",
.br
"tengo" : "10 MXN",
.br
"manzanas":{
.RS
"tengo" : 10,
.br
"colores":[ \fR/*array/dic con lista de cadenas, números o arrays.*/\fB
.RS
"verdes",
.br
"rojas"
.RE
]
.RE
}
.RE
}
.br
poner("Ganamos " .. datos.gane)
.br
poner("Gastamos " .. datos.gaste)
.br
poner("Y nos quedan " .. datos.tengo)
.br
poner("De manzanas, tenemos " .. datos.manzanas.tengo)
.br
desde (i=0; i<dic.longitud(datos.manzanas.colores); i++)
.br
.RS
poner("Color "..i..": "..datos.manzanas.colores[i]) \fR/* Obtiene colores recorriendo lista*/\fB
.RE
.br
fin
\fR
.RE
.SH Copyright
Se concede permiso, de forma gratuita, a cualquier persona que obtenga una copia
.br
de este software y archivos de documentacion asociados (el "Software"), para hacer frente
.br
el Software sin restriccion, incluyendo sin limitacion, los derechos
.br
para usar, copiar, modificar, fusionar, publicar, distribuir, sublicenciar y / o vender
.br
las copias del Software y permitir a las personas a las que el Software es
.br
amueblado para hacerlo, con sujecion a las siguientes condiciones:
.PP
El aviso de copyright anterior y este aviso de permiso se incluiran en
.br
todas las copias o partes sustanciales del Software.
.PP
EL SOFTWARE SE PROPORCIONA "TAL CUAL", SIN GARANTIA DE NINGUN TIPO, EXPRESA O
.br
IMPLICITAS, INCLUYENDO PERO NO LIMITADO A LAS GARANTIAS DE COMERCIALIZACION,
.br
IDONEIDAD PARA UN PROPOSITO PARTICULAR Y NO INFRACCION. EN NINGUN CASO EL
.br
AUTORES O TITULARES DEL COPYRIGHT SERAN RESPONSABLES POR NINGUN RECLAMO, DAÑO U OTRA
.br
RESPONSABILIDAD, YA SEA EN UNA ACCION DE CONTRATO, AGRAVIO O CUALQUIER OTRA FORMA, DERIVADOS DE,
.br
DE O EN RELACION CON EL SOFTWARE O EL USO U OTROS TRATOS EN
.br
EL SOFTWARE.


history.txt
ejecutar_pipe("echo nsns")
ejecutar_pipe("echo nsn\\s")
ejecutar_pipe("echo nsn\\\\s")
ejecutar_pipe("echo '\\\\'")
ejecutar_pipe("echo '\\\\' | sed 's/\\\\/\\\\\\\\g'")
ejecutar_pipe("echo '\\\\' | sed 'g/\\\\/\\\\\\\\s'")
ejecutar_pipe("echo '\\\\' | sed 'g/\\\\/\\\\\\\\/s'")
ejecutar_pipe("echo '\\\\' | sed 's/\\\\/\\\\\\\\/g'")
peticion("http://google.com")
peticion("https://google.com")
peticion("https://www.google.com")


src/latino.c
/*
The MIT License (MIT)
Copyright (c) Latino - Lenguaje de Programacion
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
 */

#include "latino.h"
#include "latast.h"
#include "latmem.h"
#include "latmv.h"
#include "latrepl.h"

/*
Para depurar en visual studio:
Menu propiedades del proyecto-> Debugging -> Command Arguments. Agregar
$(SolutionDir)..\ejemplos\debug.lat
----------------------------------
Para depurar en Netbeans ir a propiedades del proyecto -> Run command y Agregar
"${OUTPUT_PATH}" ejemplos/debug.lat
 */

int yydebug = 0; /* 1 para debuguear analizador lexico/sintactico */
int parse_silent = 0;
char *filename = NULL;

void str_ejecutar(lat_mv *mv);

/**
 * Muestra la version de latino en la consola
 */
static void lat_version() { printf("%s\n", LAT_DERECHOS); }

/**
 * Para crear el logo en ascii
 */
static void lat_logo() { printf("%s\n", LAT_LOGO); }

/**
 * Muestra la ayuda en la consola
 */
static void lat_ayuda() {
    lat_logo();
    lat_version();
    printf("%s\n", "Uso de latino: latino [opcion] [archivo]");
    printf("\n");
    printf("%s\n", "Opciones:");
    printf("%s\n", "-a           : Muestra la ayuda de Latino");
    printf("%s\n", "-e           : Ejecuta una cadena de codigo");
    printf("%s\n", "-v           : Muestra la version de Latino");
    printf("%s\n", "archivo      : Nombre del archivo con extension .lat");
    printf("%s\n", "Ctrl-C       : Para cerrar");
    printf("\n");
    printf("%s\n", "Variables de entorno:");
    printf("%s\n", "_____________________");
    printf("%s%s\n", "LATINO_PATH  : ", getenv("LATINO_PATH"));
    printf("%s%s\n", "LATINO_LIB   : ", getenv("LATINO_LIB"));
    printf("%s%s\n", "LATINO_LIBC   : ", getenv("LATINO_LIBC"));
    printf("%s%s\n", "HOME         : ", getenv("HOME"));
}

int main(int argc, char *argv[]) {
    setlocale(LC_ALL, "");
    setlocale(LC_CTYPE, "");
    /* para numeros decimales */
    // setlocale(LC_NUMERIC, "es_MX");
    // setlocale(LC_MONETARY, "es_MX");
    int i;
    char *infile = NULL;
    int pe = false;
    int pf = false;
    /* procesa params de latino */
    for (i = 1; i < argc; i++) {
        if (strstr(argv[1], ".lat") || argv[1][0] != '-') {
            pf = true;
            break;
        }
        if (!strcmp(argv[i], "-v")) {
            lat_version();
            return EXIT_SUCCESS;
        } else if (!strcmp(argv[i], "-a") || !strcmp(argv[i], "--help")) {
            lat_ayuda();
            return EXIT_SUCCESS;
        } else if (!strcmp(argv[i], "-e")) {
            pe = true;
        }
    }
    lat_mv *mv = latC_crear_mv();
    if (pe) {
        /* Ejecuta una cadena de codigo */
        if (argc != 3) {
            printf("Error: Se requiere una cadena para ejecución.\n");
            return EXIT_FAILURE;
        }
        char *cmd = malloc(MAX_INPUT_SIZE);
        strcpy(cmd, argv[2]);
        latC_apilar(mv, latC_crear_cadena(mv, cmd));
        str_ejecutar(mv);
        free(cmd);
    } else if (argc > 1 && pf) {
        /* Ejecuta un archivo de codigo */
        infile = argv[1];
        mv->nombre_archivo = infile;
        mv->global->REPL = false;
        mv->global->argc = argc - 1;
        int i;
        for (i = 1; i < argc; i++) {
            latL_agregar(mv, latC_checar_lista(mv, mv->global->argv),
                         latC_crear_cadena(mv, argv[i]));
        }
        int status;
        ast *nodo = latA_analizar_arch(infile, &status);
        if (status == 0 && nodo != NULL) {
            lat_objeto *mainFunc = latC_analizar(mv, nodo);
            if (mv->global->menu) {
                // instrucciones para llamar a menu
                lat_funcion *fval = (lat_funcion *)mainFunc->val.gc;
                lat_bytecode *codigo = (lat_bytecode *)fval->codigo;
                int ninst = mainFunc->ninst;
                lat_bytecode *bc = (lat_bytecode *)latM_asignar(
                    NULL, sizeof(lat_bytecode) * (ninst + 2));
                memcpy(bc, codigo, latM_tamanio(bc));
                bc[ninst - 1] = latMV_bytecode_crear(
                    LOAD_NAME, 0, 0, latC_crear_cadena(mv, "menu"), 0, 0,
                    mv->nombre_archivo);
                bc[ninst] = latMV_bytecode_crear(CALL_FUNCTION, 2, 0, NULL, 0,
                                                 0, mv->nombre_archivo);
                bc[ninst + 1] = latMV_bytecode_crear(HALT, 0, 0, NULL, 0, 0,
                                                     mv->nombre_archivo);
                lat_objeto *num = latC_crear_numerico(mv, mv->global->argc);
                latC_apilar(mv, num);
                latC_apilar(mv, mv->global->argv);
                lat_objeto *newMain = latC_crear_funcion(mv, bc, ninst + 2);
                newMain->es_vararg = 0;
                newMain->nparams = 2;
                status = latC_llamar_funcion(mv, newMain);
                latO_destruir(mv, newMain);
            } else {
                status = latC_llamar_funcion(mv, mainFunc);
                // printf("main status: %i\n", status);
                latO_destruir(mv, mainFunc);
            }
        }
        latA_destruir(nodo);
    } else {
        lat_version();
        latR_REPL(mv);
    }
    // latC_destruir_mv(mv);
    return EXIT_SUCCESS;
}
